#' @title Run CoNGA
#'
#' @description Runs CoNGA on data generated by Rdiscvr::CreateMergedTcrClonotypeFile
#' @param variable_features_file textfile with list of most variable genes
#' @param tcr_datafile Merged TCR clonotype file
#' @param gex_datafile h5 file
#' @param organism 'human' or 'rhesus'
#' @param outfile_prefix prefix to output files
#' @param gex_datatype "10x_h5"
#' @param clones_file clones file generated by CoNGA
#' @param outfile_prefix_for_qc_plots prefix to qc output files
#'
#' @export

RunCoNGA <- function(variable_features_file, tcr_datafile, gex_datafile, organism,
                      outfile_prefix, gex_datatype, clones_file, 
                      outfile_prefix_for_qc_plots) {

  if (!reticulate::py_available(initialize = TRUE)) {
    stop(paste0('Python/reticulate not configured. Run "reticulate::py_config()" to initialize python'))
  }

  if (!reticulate::py_module_available('conga')) {
    stop('The conga python package has not been installed! If you believe it has been installed, run reticulate::import("conga") to get more information and debug')
  }
  print('config complete')
  #str <- readr::read_file(system.file("scripts/run_CoNGA.py", package = "CellMembrane"))
  str <- readr::read_file("/Users/mcelfreg/scRNASeq/bimber-lab/CellMembrane/inst/scripts/run_CoNGA.py")
  print("script accessed")
  script <- tempfile()
  readr::write_file(str, script)
  print("script written")
  newstr <- paste0("run_CoNGA(features_file = '", variable_features_file,
                   "',tcr_datafile = '", tcr_datafile, "', gex_datafile = '",
                   gex_datafile, "', organism = '", organism,
                   "', outfile_prefix = '", outfile_prefix,"', gex_datatype = '",
                   gex_datatype, "', clones_file = '", clones_file,
                   "', outfile_prefix_for_qc_plots = '", 
                   outfile_prefix_for_qc_plots,"')")
  print(newstr)
  readr::write_file(newstr, script, append = TRUE)
  print("wrote script")
  system2(reticulate::py_exe(), script)
  print("finished script")
}

#' @title SeuratToCoNGA
#' @param seuratObj The seurat object
#' @param clonotypesFile The 10x clonotypes file for this run
#' @param outputDir The local path to write the output files
#' @param assayName The name of the gene expression data assay
#' @description A wrapper function to prepare a Seurat object for Conga.
#' @export
SeuratToCoNGA <- function(seuratObj, clonotypesFile, outputDir, assayName = 'RNA') {
  if (!dir.exists(outputDir)) {
    dir.create(outputDir, recursive = T)
  }
  write.table(Seurat::VariableFeatures(seuratObj), paste0(outputDir, "/varfeats.csv"), row.names = FALSE, col.names = FALSE)
  file.copy(clonotypesFile, paste0(outputDir, "/TCRs.csv"))

  DropletUtils::write10xCounts(x = seuratObj@assays[[assayName]]@counts, path = paste0(outputDir, "/GEX.h5"))
}

#' @title Run CoNGA on Seurat data and subsets
#'
#' @description Runs CoNGA on combo data and on combo data split apart into individual samples
#' @param seuratObj input Seurat object
#' @param splitFields a vector of metadata columns used for splitting the Seurat object
#' @param outputdir file path of directory for output files
#' @param organism 'human' or 'rhesus'
#' @param gex_datatype "10x_h5"
#' @param assayName The name of the gene expression data assay
#'
#' @export
runCoNGA_on_seuratdata_and_subsets <-function(seuratObj, splitFields,
                                              outputdir, organism, gex_datatype, assayName = 'RNA') {

  outputdir <- normalizePath(outputdir, mustWork = FALSE)
  if (!endsWith(outputdir, '/')) {
    outputdir <- paste0(outputdir, '/')
  }

  if (!dir.exists(outputdir)) {
    dir.create(outputdir)
  }

  maindir <- paste0(outputdir, "main/")
  if (dir.exists(maindir)) {
    print(paste0('Deleting pre-existing folder: ', maindir))
    unlink(maindir, recursive = TRUE)
  }
  dir.create(maindir)

  SeuratToCoNGA(seuratObj, maindir)
  clonesfile <- paste0(maindir, "TCRs.csv")
  clones <- read.csv(clonesfile)

  meta <- seuratObj@meta.data
  clones_sub <- subset(clones, subset = barcode %in% rownames(meta))
  write.csv(clones_sub, paste0(maindir, "TCRs_sub.csv"), row.names = F)

  grouplist <- list()
  for (i in 1:length(splitFields)) {
    grouplist[[i]] <- meta[,splitFields[i]]
  }
  splitlist <- split(meta, grouplist)
  barcodelist <- lapply(splitlist, rownames)
  
  variable_features_file <- paste0(maindir, "varfeats.csv")
  outfile_prefix <- maindir
  clones_file <- paste0(maindir, "clones.tsv")
  tcr_datafile <- paste0(maindir, "TCRs_sub.csv")
  gex_datafile <- paste0(maindir, "GEX.h5")
  outfile_prefix_for_qc_plots <- paste0(maindir, "QC_plots/")
  RunCoNGA(variable_features_file, tcr_datafile, gex_datafile, organism,
           outfile_prefix, gex_datatype, clones_file,
           outfile_prefix_for_qc_plots)
  
  for (splitname in names(barcodelist)) {
    if (length(barcodelist[[splitname]]) > 0){
      current_dir <- paste0(outputdir, splitname, "/")
      dir.create(current_dir)
      clones_sub <- subset(clones, subset = barcode %in% barcodelist[[splitname]])
      tcr_datafile <- paste0(current_dir, "TCRs_sub.csv")
      write.csv(clones_sub, tcr_datafile, row.names = F)
      gex_datafile <- paste0(current_dir, "GEX.h5")
      seuratObj_sub <- subset(seuratObj, cells = barcodelist[[splitname]])
      seuratObj_sub <- NormalizeAndScale(seuratObj_sub)
      DropletUtils::write10xCounts(x = seuratObj_sub@assays[[assayName]]@counts, path = gex_datafile)
      variable_features_file <- paste0(current_dir, "varfeats.csv")
      write.table(VariableFeatures(seuratObj_sub), variable_features_file, row.names = FALSE, col.names = FALSE)
      
      outfile_prefix <- current_dir
      clones_file <- paste0(current_dir, "clones.tsv")
      outfile_prefix_for_qc_plots <- paste0(current_dir, "QC_plots/")
      RunCoNGA(variable_features_file, tcr_datafile, gex_datafile, organism,
               outfile_prefix, gex_datatype, clones_file,
               outfile_prefix_for_qc_plots)
    }
  }
  
  clones_df <- read.csv(paste0(maindir, "/TCRs_sub.csv"))
  clones <- (clones_df$raw_clonotype_id)
  names(clones) <- clones_df$barcode
  seuratObj <- Seurat::AddMetaData(seuratObj, clones, col.name = "clone")
  
  df <- data.frame()
  for (sample in list.files(outputdir)[list.files(outputdir) != "main"]) {
    sample_obs <- paste0(outputdir, "/", sample, "/adata2obs.csv")
    tempdf <- read.csv(sample_obs, header=TRUE, row.names = 1)
    tempdf$sample <- sample
    df <- rbind(df, tempdf)
  }
  dfout <- df %>% select(c(clusters_gex, clusters_tcr, nndists_gex, nndists_tcr,
                           is_invariant, conga_scores, conga_fdr_values, sample))
  colnames(dfout) <- paste0("ind_", colnames(dfout))
  seuratObj <- Seurat::AddMetaData(seuratObj, dfout)
  
  combo_df <- read.csv(paste0(outputdir, "/main/adata2obs.csv"), header=TRUE, row.names = 1)
  combo_dfout <- combo_df %>% select(c(clusters_gex, clusters_tcr, nndists_gex, nndists_tcr,
                                       is_invariant, conga_scores, conga_fdr_values))
  colnames(combo_dfout) <- paste0("combo_", colnames(combo_dfout))
  seuratObj <- Seurat::AddMetaData(seuratObj, combo_dfout)

  return(seuratObj)
}
